processing/src/main/java/org/apache/druid/segment/CursorMaker.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment;
+
+import org.apache.druid.java.util.common.UOE;
+import org.apache.druid.segment.vector.VectorCursor;
+
+import javax.annotation.Nullable;
+import java.io.Closeable;
+
+public interface CursorMaker extends Closeable
```

Comment:
I wonder if `CursorHolder` is a better name than `CursorMaker`? Using "holder" rather than "maker" better emphasizes the fact that the `CursorMaker` (or `CursorHolder`) is closeable— it suggests it's a container that wraps a `Cursor` or `VectorCursor` + the resources used by the cursor. It also avoids the awkward name `CursorMakerFactory`.

If we did that, I think we'd also change:

- `makeCursor` -> `asCursor()` or `getCursor()`
- `makeVectorCursor` -> `asVectorCursor()` or `getVectorCursor()`
- `CursorMakerFactory` -> `CursorHolderFactory` or even stick with `CursorFactory`
- `CursorMakerFactory#asCursorMaker` -> `#makeCursorHolder`
- possibly update `CursorMaker` (`CursorHolder`) implementations to stash the cursor in a local field, so `asCursor()` or `getCursor()` could be called multiple times and return the same cursor. This would enable query engines to pass around the `CursorHolder` as a unified `Closeable` + holder for the actual cursor, which would potentially simplify the logic around tracking that things are closed properly.

Reply:
sounds good to me, `CursorMaker` was the name before I reworked things to allow it to be closeable instead of the strange mixed model of the old thing where `Sequence<Cursor>` closed itself by consuming the sequence and `VectorCursor` was closeable.

Reply:
changed the names, I haven't yet switched the `CursorHolder` implementations to stash cursor/vector cursor yet, though looking into it